jax/core.py

```diff
@@ -888,17 +889,73 @@ def maybe_find_leaked_tracers(x: Optional[Union[MainTrace, Sublevel]]
   """
   if not getattr(threading.current_thread(), 'pydev_do_not_trace', True):
     warnings.warn(TRACER_LEAK_DEBUGGER_WARNING)
-  # Trigger garbage collection to filter out cyclical dependency false positives
+  # Trigger garbage collection to filter out unreachable objects that are alive
+  # only due to cyclical dependencies. (We don't care about unreachable leaked
+  # tracers since they can't interact with user code and cause a problem.)
   gc.collect()
   traces = list(filter(lambda x: isinstance(x, Trace), gc.get_referrers(x)))
   tracers = list(filter(lambda x: isinstance(x, Tracer), gc.get_referrers(*traces)))
   return tracers
 
 def leaked_tracer_error(name: str, t, tracers: List[Tracer]) -> Exception:
   assert tracers
-  msgs = '\n\n'.join(f'{tracer}{tracer._origin_msg()}' for tracer in tracers)
+  why = partial(_why_alive, {id(tracers)})
+  msgs = '\n\n'.join(f'{tracers[i]}{tracers[i]._origin_msg()}{why(tracers[i])}'
+                     for i in range(len(tracers)))
   return Exception(f'Leaked {name} {t}. Leaked tracer(s):\n\n{msgs}\n')
 
+def _why_alive(ignore_ids: Set[int], x: Any) -> str:
+  parents = lambda x: [r for r in gc.get_referrers(x) if id(r) not in ignore_ids]
+  child, lines, seen = x, [], set()
+  while (id(child) not in seen and type(child) is not types.ModuleType
+         and parents(child)):
+    parent = parents(child)[0]  # just pick one parent
+
+    # special-case namespaces and closures, handled in _why_alive_container_info
+    # (if this heuristic causes problems, this code block can be commented out).
+    # TODO(mattjj): after Python 3.7 is unsupported, replace with types.CellType
+    cell_type = type((lambda x: lambda: x)(10.28).__closure__[0]) # type: ignore
+    if (isinstance(parent, dict) and
+        getattr(parents(parent)[0], '__dict__', None) is parents(child)[0]):
+      parent = parents(parent)[0]
+    elif type(parent) is cell_type:
+      parent = parents(parents(parent)[0])[0]
```

Comment:
Yes, that's the idea: a little shortcut. Without this we get this printed:

```
<DynamicJaxprTracer 139764213674128> is referred to by <cell 139764213646192>
<cell 139764213646192> is referred to by <tuple 139764213646096>[0]
<tuple 139764213646096> is referred to by <function 139764213640064>
<function 139764213640064> is referred to by <dict 139766096500800>['hi']
<dict 139766096500800> is referred to by <A 139766089094432>.dct
<A 139766089094432> is referred to by __main__.a
```

but with it we get:

```
<DynamicJaxprTracer 140139354034480> is referred to by <function 140139405110576> (foo) closed-over variable x
<function 140139405110576> is referred to by <dict 140141242837056>['hi']
<dict 140141242837056> is referred to by <A 140141237605664>.dct
<A 140141237605664> is referred to by __main__.a
```

I'll add a comment.

Reply:
```
    # For namespaces (like modules and class instances) and closures, the
    # references may form a simple chain: e.g. instance refers to its own
    # __dict__ which refers to child, or function refers to its __closure__
    # which refers to cells which refer to child. In these cases, we can provide
    # a more intuitive description by collapsing the chain into a single
    # parent->child jump. We do that by setting `parent` here to be a
    # grandparent (or great-grandparent) of `child`, and then handling that case
    # in _why_alive_container_info.
    # To prevent this collapsing behavior, just comment out this code block.
```